File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
const productExceptSelf = nums => {
    
    let ans = new Array(nums.length);
    
    ans[0] = 1;
    
    for (let i = 1; i < ans.length; i++) {
        ans[i] = ans[i - 1] * nums[i - 1];
    }
    
    let right = 1;    

    for (let i = ans.length - 1; i >= 0; i--) {
        ans[i] *= right;
        right *= nums[i];
    }
    
    return ans;
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        leftSide, rightSide = [1 for i in range(len(nums))], [1 for i in range(len(nums))]
        for i in range(len(nums)):
            if i != 0:
                leftSide[i] = leftSide[i - 1] * nums[i - 1]
        for i in range(len(nums) - 1, -1, -1):
            if i != len(nums) - 1:
                rightSide[i] = rightSide[i + 1] * nums[i + 1]
        return [leftSide[i] * rightSide[i] for i in range(len(nums))]
            